File: 661_Image_Smoother.java
/*
https://leetcode.com/problems/image-smoother/
Given a 2D integer matrix M representing the gray scale of an image, you need to design a smoother to make the gray
scale of each cell becomes the average gray scale (rounding down) of all the 8 surrounding cells and itself. If a cell
has less than 8 surrounding cells, then use as many as you can.

Iterate over cells of image. For each cell count neighbours and sum their values. Set smoothed pixel to sum / count.
Time - O(mn)
Space - O(mn)
*/

class Solution {
    public int[][] imageSmoother(int[][] M) {

        int rows = M.length;
        int cols = M[0].length;
        int[][] smoothed = new int[rows][cols];

        for (int r = 0; r < rows; ++r) {
            for (int c = 0; c < cols; ++c) {

                int count = 0;
                int total = 0;
                for (int nr = r - 1; nr < r + 2; ++nr) {
                    for (int nc = c - 1; nc < c + 2; ++nc) {
                        if (nr < 0 || nr >= rows || nc < 0 || nc >= cols)
                            continue;
                        ++count;
                        total += M[nr][nc];
                    }
                }
                smoothed[r][c] = total / count;
            }
        }
        return smoothed;
    }
}

File: 661_Image_Smoother.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/image-smoother/
# Given a 2D integer matrix M representing the gray scale of an image, you need to design a smoother to make the gray
# scale of each cell becomes the average gray scale (rounding down) of all the 8 surrounding cells and itself.
# If a cell has less than 8 surrounding cells, then use as many as you can.

# Iterate over M. For each cell, count and sum the neighbours (including self).
# Time - O(mn)
# Space - O(mn)

class Solution(object):
    def imageSmoother(self, M):
        """
        :type M: List[List[int]]
        :rtype: List[List[int]]
        """
        rows, cols = len(M), len(M[0])
        smoothed = [[0 for _ in range(cols)] for _ in range(rows)]

        for r in range(rows):
            for c in range(cols):

                nbors, total = 0, 0
                for dr in (-1, 0, 1):
                    for dc in (-1, 0, 1):

                        if r + dr < 0 or r + dr >= rows or c + dc < 0 or c + dc >= cols:
                            continue
                        total += M[r + dr][c + dc]
                        nbors += 1

                smoothed[r][c] = total // nbors         # round down

        return smoothed